Core/include/Acts/TrackFinder/CombinatorialKalmanFilter.hpp

```diff
@@ -0,0 +1,1255 @@
+
+          // Screen out material effects info
+          ACTS_VERBOSE("Material effects on surface: "
+                       << surface->geoID()
+                       << " at update stage: " << updateStage << " are :");
+          ACTS_VERBOSE("eLoss = "
+                       << interaction.Eloss << ", "
+                       << "variancePhi = " << interaction.variancePhi << ", "
+                       << "varianceTheta = " << interaction.varianceTheta
+                       << ", "
+                       << "varianceQoverP = " << interaction.varianceQoverP);
+
+          // Update the state and stepper with material effects
+          interaction.updateState(state, stepper);
+        }
+      }
+
+      if (not hasMaterial) {
+        // Screen out message
+        ACTS_VERBOSE("No material effects on surface: " << surface->geoID()
+                                                        << " at update stage: "
+                                                        << updateStage);
+      }
+    }
+
+    /// @brief Kalman actor operation : finalize
+    ///
+    /// @tparam propagator_state_t is the type of Propagagor state
+    /// @tparam stepper_t Type of the stepper
+    ///
+    /// @param state is the mutable propagator state object
+    /// @param stepper The stepper in use
+    /// @param result is the mutable result state object
+    template <typename propagator_state_t, typename stepper_t>
+    Result<void> finalize(propagator_state_t& state, const stepper_t& stepper,
+                          result_type& result) const {
+      // Get the index of measurement states;
+      std::vector<size_t> measurementIndices;
+      auto lastState = result.fittedStates.getTrackState(result.currentTip);
+      if (lastState.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag)) {
+        measurementIndices.push_back(result.currentTip);
+      }
+      // Count track states to be smoothed
+      size_t nStates = 0;
+      result.fittedStates.applyBackwards(result.currentTip, [&](auto st) {
+        // Smoothing will start from the last measurement state
+        if (measurementIndices.empty()) {
+          // No smoothed parameter for the last few non-measurment states
+          st.data().ismoothed = detail_lt::IndexData::kInvalid;
+        } else {
+          nStates++;
+        }
+        size_t iprevious = st.previous();
+        if (iprevious != Acts::detail_lt::IndexData::kInvalid) {
+          auto previousState = result.fittedStates.getTrackState(iprevious);
+          if (previousState.typeFlags().test(
+                  Acts::TrackStateFlag::MeasurementFlag)) {
+            measurementIndices.push_back(iprevious);
+          }
+        }
+      });
+      // Return error if the track has no measurement states (but this should
+      // not happen)
+      if (measurementIndices.empty()) {
+        return CombinatorialKalmanFilterError::SmoothFailed;
```

Comment:
Can you add a `ACTS_ERROR` here?

Reply:
Thanks. They have been added here and elsewhere.